### PR TITLE
Add bulk indexing to result index handler

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are a lot of dependencies, I only list the main class and test code to save reviewers' time. The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big one in the end and merge once after all review PRs get approved.

*Issue #, if available:*

*Description of changes:*
For multi-entities' results, we bulk indexing entities’ anomaly results of a detector instead of doing it one entity by another. Bulk requests will yield much better performance than single-document index requests.

This PR implements bulk indexing logic in MultiEntityResultHandler.

Testing done:
1. Will add unit tests.
2. Manual testing passes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/DetectionStateHandler.java [hunk 1]
```diff
@@ -100,11 +100,11 @@ public DetectionStateHandler(
             DetectorInternalState.DETECTOR_STATE_INDEX,
             createIndex,
             indexExists,
-            true,
             clientUtil,
             indexUtils,
             clusterService
         );
+        this.fixedDoc = true;
```
Comment (ohltyler) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/DetectionStateHandler.java:107]: why is `fixedDoc` true for `DetectionStateHandler`, and `false` for `AnomalyIndexHandler`?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyIndexHandler.java [hunk 2]
```diff
@@ -90,12 +101,22 @@ public AnomalyIndexHandler(
...
         this.clusterService = clusterService;
     }
 
+    /**
+     * Since the constructor needs to provide injected value and Guice does not allow Boolean to be there
+     * (claiming it does not know how to instantiate it), caller needs to manually set it to true if
+     * it want to save to a specific doc.
+     * @param fixedDoc whether to save to a specific doc Id
+     */
+    public void setFixedDoc(boolean fixedDoc) {
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyIndexHandler.java:116]: Minor: just FYI, Guice can support primitive type for injection. To do that, we need use bind with Named annotation.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/MultiEntityResultHandler.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            clientUtil,
+            indexUtils,
+            clusterService
+        );
+        this.nodeStateManager = nodeStateManager;
+        this.clock = clock;
+    }
+
+    public void write(AnomalyResult toSave, ADResultBulkRequest currentBulkRequest) {
+        currentBulkRequest.add(toSave);
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/MultiEntityResultHandler.java:89]: why do we need this simple method?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/DetectionStateHandler.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/DetectionStateHandler.java:107]: For DetectionStateHandler, we are writing to a specific doc.  By default, it is false.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyIndexHandler.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyIndexHandler.java:116]: Could you give me an example with code link?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/MultiEntityResultHandler.java (hunk 3 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/MultiEntityResultHandler.java:89]: removed

**Review by ohltyler — APPROVED** [review]

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyIndexHandler.java (hunk 2 above)
Comment (weicongs-amazon) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyIndexHandler.java:116]: similar like this: https://medium.com/@pavankumarreddy/injecting-dynamic-primitive-values-in-guice-26a1521b25b9

**Review by weicongs-amazon — APPROVED** [review]

**Pull request closed without merging**